Change default scope for `PreferenceService.set()`

colin-grant-work opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

Fixes #9152 by changing how the scope to which to apply a call to `PreferenceService.set()` is calculated. Formerly, if no scope was provided, `PreferenceScope.Folder` was used if a `URI` was supplied, `PreferenceScope.Workspace` otherwise. Now `PreferenceScope.User` is used instead of `PreferenceScope.Workspace`.

In addition, an error was previously thrown if `set` was called with a `sectionName` (`tasks`, `launch`) in User scope, but that is no longer appropriate because some section configurations are available in user scope, so now that check is delegated to `PreferenceSchema.isValidInScope()`. It could be argued that the check could be omitted, since an error will be thrown later in the pipeline (when preference changes are processed). 

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

1. 3. (You may want to set the environment variable THEIA_CONFIG_DIR to some temp dir)
2. Run the File > Auto Save menu item.
3. Observe that the setting is changed in user scope.
4. Create a command that calls e.g. `PreferenceService.set('tasks', {version: '2.0.0', tasks: []});`. E.g. add this to `PreferencesContribution.registerCommands()`:
```typescript
        commands.registerCommand({id: 'test-setting-preferences', label: 'Set a preference using PreferenceService.set'}, {
            execute: () => this.preferenceService.set('tasks', {version: '2.0.0', tasks: []}),
        });
```
5. Observe that it that setting takes effect in user scope and no error is logged.
6. Set some other preferences using the preference service and see how it goes.
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Colin Grant <colin.grant@ericsson.com>

kittaakos commented:
> Formerly, if no scope was provided, `PreferenceScope.Folder` was used if a `URI` was supplied, `PreferenceScope.Workspace` otherwise. Now `PreferenceScope.User` is used instead of `PreferenceScope.Workspace`.

I like the idea but VS Code uses workspace as the default if I am not mistaken:

> otherwise to Workspace settings.

Could you please check the [docs](https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration), @colin-grant-work. Maybe, I am overlooking something. Thank you!
> `update(section: string, value: any, configurationTarget?: ConfigurationTarget | boolean, overrideInLanguage?: boolean): Thenable<void>`
> The configuration target or a boolean value. - If true updates Global settings. - If false updates Workspace settings. - If undefined or null updates to Workspace folder settings if configuration is resource specific, otherwise to Workspace settings.

colin-grant-work commented:
> > `update(section: string, value: any, configurationTarget?: ConfigurationTarget | boolean, overrideInLanguage?: boolean): Thenable<void>`
> > The configuration target or a boolean value. - If true updates Global settings. - If false updates Workspace settings. - If undefined or null updates to Workspace folder settings if configuration is resource specific, otherwise to Workspace settings.

I agree, looks like `workspace` is the default for VSCode, though the boolean option is a weird twist (must be for backwards compatibility to an earlier system, I guess).

I'm not committed to changing the default scope. It was my understanding that what was requested in the bug report ticket was a more general change to the behavior of `.set`, rather than just setting the `editor.autoSave` preference in user scope. If we only want to fix where the auto-save preference gets set, that's an easy fix, too. I'd still take out the check for whether we're setting a `sectionName` in user scope, since that check is no longer relevant, but happy to return to the old defaulting.

vince-fugnitto commented on packages/editor/src/browser/editor-command.ts:
```diff
@@ -333,6 +333,6 @@ export class EditorCommandContribution implements CommandContribution {
     }
     private async toggleAutoSave(): Promise<void> {
-        this.preferencesService.set(EditorCommandContribution.AUTOSAVE_PREFERENCE, this.isAutoSaveOn() ? 'off' : 'on');
+        this.preferencesService.set(EditorCommandContribution.AUTOSAVE_PREFERENCE, this.isAutoSaveOn() ? 'off' : 'on', PreferenceScope.User);
```
From what I understood, the issue with setting the preference with `PreferenceScope.User` means that it will not work for the following use-case which is supported in vscode:

1. `editor.autoSave: false` in the workspace settings (`{workspace}/.theia/settings.json`).
2. attempt to toggle the `autoSave` with the command.
3. the preference value at the workspace scope is not updated while in vscode it is.

colin-grant-work replied:
Hm... so that clarifies where the problem with the current implementation really is. `editor.autoSave` should not be available in folder scope - it should only be set in a local `settings.json` in a single-root workspace, and in a multi-root workspace, only the workspace value should be picked up. If we're picking up a folder value in a multi-root workspace, that's a bug, and I'll look into that. Otherwise, it looks like setting it at workspace level is VSCode's behavior, and would avoid any awkward override situations.

vince-fugnitto replied:
@colin-grant-work I might have confused the `workspace` vs `folder` scopes, what I mean is the value under `{workspace}/.vscode/settings.json` is updated when toggling the command which does not happen in Theia if we are to specifically set the scope to `User`.

colin-grant-work replied:
Yes, if there is a bug that is making us pick up folder-scoped preferences, that's the fix to do. Then we should *not* set it in user scope.

kittaakos replied:
> From what I understood, the issue with setting the preference with `PreferenceScope.User` means that it will not work for the following use-case which is supported in vscode:
> 
> 1. `editor.autoSave: false` in the workspace settings (`{workspace}/.theia/settings.json`).
> 2. attempt to toggle the `autoSave` with the command.
> 3. the preference value  (`{workspace}/.theia/settings.json`) is not updated while in vscode it is (`.vscode`).

First, we can check if the workspace config contains `editor.autoSave`. If yes, we change the value there. If no, we change it in the user config. Can this work?
 - Users get the Code behavior via the menu or command palette.
 - If users want to modify it manually via the `settings.json`, they know what they're doing.
 - If you're an extension developer, you either want to trigger the default functionality via the command, so we are good. Or you are doing custom things via the preferences service API.

Thoughts? Am I overlooking something?

colin-grant-work replied:
So the code that actually performs this update in VSCode is not the `update` method to which you posted the docs earlier or this thread. It's the `updateValue` method, for which I don't think we have an equivalent. That code looks pretty similar to what you're describing: it loops through all scopes, finding the lowest one in which the value of the given preference isn't currently `undefined` (with user scope as default), and sets the value there. If the value is being deleted, it deletes it in all scopes in which it's defined.

https://github.com/microsoft/vscode/blob/257fbae3f34804eeb427c10bceb4c0ed95870552/src/vs/workbench/services/filesConfiguration/common/filesConfigurationService.ts#L185-L196


https://github.com/microsoft/vscode/blob/257fbae3f34804eeb427c10bceb4c0ed95870552/src/vs/workbench/services/filesConfiguration/common/filesConfigurationService.ts#L185-L196


https://github.com/microsoft/vscode/blob/257fbae3f34804eeb427c10bceb4c0ed95870552/src/vs/workbench/services/configuration/browser/configurationService.ts#L783-L808

If we want to have that logic available for this command, we can implement methods to mirror that in the `PreferenceService`, or we can just do it manually for this command.

kittaakos replied:
> we can implement methods to mirror that in the `PreferenceService`

Good idea. Let's keep the new API aligned to the `updateValue` from Code. We can use this new API for auto-save and maybe for others too (theming?).

As an extension developer, I would prefer to use an API that has the same semantics as a VS Code extension API. What do you think?

colin-grant-work replied:
I've implemented this. I'll add some tests next week to make sure it's doing what it should.

colin-grant-work replied:
> I've implemented this. I'll add some tests next week to make sure it's doing what it should.

@vince-fugnitto, the tests are still planned 😃

vince-fugnitto replied:
@colin-grant-work sorry about that :)

kittaakos commented on packages/core/src/browser/preferences/preference-service.ts:
```diff
@@ -476,6 +487,50 @@ export class PreferenceServiceImpl implements PreferenceService {
+                return inspection.workspaceValue;
+            case PreferenceScope.Folder:
+                return inspection.workspaceFolderValue;
+            default: // Should be unreachable
+                return undefined;
```
FYI: If you want to handle the _"should be unreachable"_ part the TypeScript way, you have to use `never`, and you will get a compile-time error if you add a new enum:
```ts
    protected getScopedValueFromInspection<T>(inspection: PreferenceInspection<T>, scope: PreferenceScope): T | undefined {
        const handleUnexpectedScope: (never: never) => never = () => {
            throw new Error('Should be unreachable');
        };
        switch (scope) {
            case PreferenceScope.Default:
                return inspection.defaultValue;
            case PreferenceScope.User:
                return inspection.globalValue;
            case PreferenceScope.Workspace:
                return inspection.workspaceValue;
            case PreferenceScope.Folder:
                return inspection.workspaceFolderValue;
            default: // Should be unreachable
                return handleUnexpectedScope(scope);
        }
    }
```
You do not have to change anything, though 😊 I just wanted to bring this up; please ignore if you already know this.

colin-grant-work replied:
Very nice - good to make the person who changes the enum pay 😅

kittaakos reviewed (approved):
I tried it; it works very well. The code looks great. Thank you so much, @colin-grant-work 👍

vince-fugnitto reviewed (commented):
@colin-grant-work I believe the tests will need adjusting, they currently fail when testing the preferences.

colin-grant-work commented:
@vince-fugnitto, I've implemented some tests here and modified the tests that were failing. If you wouldn't mind taking a look, I think it should be just about good to go.

vince-fugnitto reviewed (approved):
The changes look good to me, and I confirmed that `autoSave` now behaves like expected.

I only have a minor comment about the `api-test` where the import has an error but does not affect functionality.

colin-grant-work commented on examples/api-tests/src/launch-preferences.spec.js:
```diff
@@ -30,7 +30,7 @@ describe('Launch Preferences', function () {
     const { assert } = chai;
 
-    const { PreferenceService, PreferenceScope } = require('@theia/core/lib/browser/preferences/preference-service');
+    const { PreferenceService, PreferenceScope, PreferenceInspection } = require('@theia/core/lib/browser/preferences/preference-service');
```
The behavior there seems to be a bit inconsistent between environments, but I've modified the code in a way that seems to make the IDE happy in all of the environments available to me.
